examples/dqn/dqn_cartpole.py

```diff
@@ -0,0 +1,183 @@
+        storage=LazyTensorStorage(
+            max_size=cfg.buffer.buffer_size,
+            device=device,
+        ),
+        batch_size=cfg.buffer.batch_size,
+    )
+
+    # Create the loss module
+    loss_module = DQNLoss(
+        value_network=model,
+        gamma=cfg.loss.gamma,
+        loss_function="l2",
+        delay_value=True,
+    )
+    loss_module.make_value_estimator(gamma=cfg.loss.gamma)
+    target_net_updater = HardUpdate(
+        loss_module, value_network_update_interval=cfg.loss.hard_update_freq
+    )
+
+    # Create the optimizer
+    optimizer = torch.optim.Adam(loss_module.parameters(), lr=cfg.optim.lr)
+
+    # Create the logger
+    logger = None
+    if cfg.logger.backend:
+        exp_name = generate_exp_name("DQN", f"CartPole_{cfg.env.env_name}")
+        logger = get_logger(cfg.logger.backend, logger_name="dqn", experiment_name=exp_name)
+
+    # Create the test environment
+    test_env = make_env(cfg.env.env_name, device)
+
+    # Main loop
+    collected_frames = 0
+    start_time = time.time()
+    num_updates = cfg.loss.num_updates
+    batch_size = cfg.buffer.batch_size
+    test_interval = cfg.logger.test_interval
+    num_test_episodes = cfg.logger.num_test_episodes
+    frames_per_batch = cfg.collector.frames_per_batch
+    pbar = tqdm.tqdm(total=cfg.collector.total_frames)
+    sampling_start = time.time()
+
+    for data in collector:
+
+        log_info = {}
+        sampling_time = time.time() - sampling_start
+        pbar.update(data.numel())
+        data = data.reshape(-1)
+        current_frames = data.numel()
+        replay_buffer.extend(data.to(device))
+        collected_frames += current_frames
+        model_explore.step(current_frames)
+
+        # Get training rewards, episode lengths and q-values
+        log_info.update(
+            {
+                "train/q_values": (data["action_value"] * data["action"]).sum().item()
+                / frames_per_batch,
+            }
+        )
+        episode_rewards = data["next", "episode_reward"][data["next", "done"]]
+        if len(episode_rewards) > 0:
+            episode_length = data["next", "step_count"][data["next", "done"]]
+            log_info.update(
+                {
+                    "train/episode_reward": episode_rewards.mean().item(),
+                    "train/episode_length": episode_length.sum().item()
+                    / len(episode_length),
+                }
+            )
+
+        # optimization steps
+        q_losses = TensorDict({}, batch_size=[num_updates])
+        training_start = time.time()
+        for j in range(num_updates):
+            sampled_tensordict = replay_buffer.sample(batch_size)
+            loss_td = loss_module(sampled_tensordict)
+            q_loss = loss_td["loss"]
+            optimizer.zero_grad()
+            q_loss.backward()
+            optimizer.step()
+            target_net_updater.step()
+            q_losses[j] = loss_td.select("loss").detach()
```

Comment:
Could/should we log more things?

Reply:
loss_td only contains "loss"

Reply:
haha ok then it's both concise and complete :)